### PR TITLE
Add upload_store_sales_transactions.pl

### PR DESCRIPTION
**Pull request by wayneluzw**
Following Java Golden CL at https://github.com/googleads/google-ads-java/pull/305/files

### CODE REVIEW

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+use strict;
+use warnings;
+use utf8;
+
+use FindBin qw($Bin);
+use lib "$Bin/../../lib";
+
+use Google::Ads::GoogleAds::Client;
+use Google::Ads::GoogleAds::Utils::GoogleAdsHelper;
+
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:31]: nit: better to remove this blank line for consistency.

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+use Google::Ads::GoogleAds::V3::Resources::OfflineUserDataJob;
+use Google::Ads::GoogleAds::V3::Common::OfflineUserAddressInfo;
+use Google::Ads::GoogleAds::V3::Common::StoreSalesMetadata;
+use Google::Ads::GoogleAds::V3::Common::StoreSalesThirdPartyMetadata;
+use Google::Ads::GoogleAds::V3::Common::TransactionAttribute;
+use Google::Ads::GoogleAds::V3::Common::UserData;
+use Google::Ads::GoogleAds::V3::Common::UserIdentifier;
+use Google::Ads::GoogleAds::V3::Enums::OfflineUserDataJobTypeEnum
+  qw(STORE_SALES_UPLOAD_FIRST_PARTY STORE_SALES_UPLOAD_THIRD_PARTY);
+use Google::Ads::GoogleAds::V3::Errors::GoogleAdsError;
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:41]: no need to import errors & responses as they're not used for constructions.

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+use
+  Google::Ads::GoogleAds::V3::Services::OfflineUserDataJobService::AddOfflineUserDataJobOperationsResponse;
+use
+  Google::Ads::GoogleAds::V3::Services::OfflineUserDataJobService::CreateOfflineUserDataJobResponse;
+use
+  Google::Ads::GoogleAds::V3::Services::OfflineUserDataJobService::OfflineUserDataJobOperation;
+use Google::Ads::GoogleAds::V3::Services::OfflineUserDataJobService;
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:48]: No need to import services, cause not used for construction.

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+  my (
+    $api_client,                 $customer_id,
+    $offline_user_data_job_type, $conversion_action_id,
+    $external_id,                $advertiser_upload_date_time,
+    $bridge_map_version_id,      $partner_id
+  ) = @_;
+
+  my $offline_user_data_job_service = $api_client->OfflineUserDataJobService();
+
+  # Creates an offline user data job for uploading transactions.
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:95]: Please use first-person verb "Create" as Perl convention as in-method comment. 
Followed the convention from WebAPII.

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+  my $offline_user_data_job_resource_name = create_offline_user_data_job(
+    $offline_user_data_job_service, $customer_id,
+    $offline_user_data_job_type,    $external_id,
+    $advertiser_upload_date_time,   $bridge_map_version_id,
+    $partner_id
+  );
+
+  # Adds transactions to the job.
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:103]: Add

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+
+  # TIP: If you are migrating from the AdWords API, please note that Google Ads
+  # API uses the term "fraction" instead of "rate". For example, loyaltyRate in
+  # the AdWords API is called loyaltyFraction in the Google Ads API.
+
+  my $store_sales_metadata =
+    # Please refer to https://support.google.com/google-ads/answer/7506124 for
+    # additional details.
+    Google::Ads::GoogleAds::V3::Common::StoreSalesMetadata->new({
+      # Sets the fraction of your overall sales that you (or the advertiser,
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:156]: Set

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+      # in the third party case) can associate with a customer (email, phone
+      # number, address, etc.) in your database or loyalty program.
+      # For example, set this to 0.7 if you have 100 transactions over 30
+      # days, and out of those 100 transactions, you can identify 70 by an
+      # email address or phone number.
+      loyaltyFraction => 0.7,
+      # Sets the fraction of sales you're uploading out of the overall sales
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:163]: Set

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+      # that you (or the advertiser, in the third party case) can associate
+      # with a customer. In most cases, you will set this to 1.0.
+      # Continuing the example above for loyalty fraction, a value of 1.0 here
+      # indicates that you are uploading all 70 of the transactions that can
+      # be identified by an email address or phone number.
+      transactionUploadFraction => 1.0
+    });
+
+  if ($offline_user_data_job_type eq STORE_SALES_UPLOAD_THIRD_PARTY) {
+    # Creates additional metadata required for uploading third party data.
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:173]: Create

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+    my $store_sales_third_party_metadata =
+      Google::Ads::GoogleAds::V3::Common::StoreSalesThirdPartyMetadata->new({
+        # The date/time must be in the format "yyyy-MM-dd hh:mm:ss".
+        advertiserUploadDateTime => $advertiser_upload_date_time,
+
+        # Sets the fraction of transactions you received from the advertiser
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:179]: Set

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+        # that have valid formatting and values. This captures any transactions
+        # the advertiser provided to you but which you are unable to upload to
+        # Google due to formatting errors or missing data.
+        # In most cases, you will set this to 1.0.
+        validTransactionFraction => 1.0,
+        # Sets the fraction of valid transactions (as defined above) you
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:185]: Set

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+        # received from the advertiser that you (the third party) have matched
+        # to an external user ID on your side.
+        # In most cases, you will set this to 1.0.
+        partnerMatchFraction => 1.0,
+
+        # Sets the fraction of transactions you (the third party) are uploading
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:191]: Set

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+        # fraction above.
+        # 2. You matched to an external user ID on your side. See partner match
+        # fraction above.
+        # In most cases, you will set this to 1.0.
+        partnerUploadFraction => 1.0,
+
+        # Please speak with your Google representative to get the values to use
+        # for the bridge map version and partner IDs.
+
+        # Sets the version of partner IDs to be used for uploads.
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:204]: Set

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+        bridgeMapVersionId => $bridge_map_version_id,
+        # Sets the third party partner ID uploading the transactions.
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:206]: Set

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+
+  # Issue a request to add the operations to the offline user data job.
+  my $response = $offline_user_data_job_service->add_operations({
+      resourceName         => $offline_user_data_job_resource_name,
+      enablePartialFailure => "true",
+      operations           => build_offline_user_data_job_operations(
+        $customer_id, $conversion_action_id
+      )});
+
+  # Prints the status message if any partial failure error is returned.
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:250]: Print

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+              })})
+      ],
+      transactionAttribute =>
+        Google::Ads::GoogleAds::V3::Common::TransactionAttribute->new({
+          conversionAction =>
+            Google::Ads::GoogleAds::V3::Utils::ResourceNames::conversion_action(
+            $customer_id, $conversion_action_id
+            ),
+          currencyCode => "USD",
+          # Converts the transaction amount from $200 USD to micros.
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:288]: Convert

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+          transactionAmountMicros => 200000000,
+          # Specifies the date and time of the transaction. This date and time
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:290]: Specify

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+      ],
+      transactionAttribute =>
+        Google::Ads::GoogleAds::V3::Common::TransactionAttribute->new({
+          conversionAction =>
+            Google::Ads::GoogleAds::V3::Utils::ResourceNames::conversion_action(
+            $customer_id,
+            $conversion_action_id
+            ),
+          currencyCode => "EUR",
+          # Converts the transaction amount from 450 EUR to micros.
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:320]: Convert

File: examples/remarketing/upload_store_sales_transactions.pl [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+          transactionAmountMicros => 450000000,
+          # Specifies the date and time of the transaction. This date and time
```
Comment (wfansh) [inline examples/remarketing/upload_store_sales_transactions.pl:322]: Specify

File: examples/remarketing/upload_store_sales_transactions.pl (hunk 3 above)
Comment (wayneluzw) [reply examples/remarketing/upload_store_sales_transactions.pl]: OfflineUserDataJobOperation is used for construction but the rest are unnecessary.

**Pull request merged**